### PR TITLE
Collateral increases in value and should be able to be withdrawn

### PR DESCRIPTION
**Pull request by Jouzo**
#### What kind of PR is this?:

/kind fix

#### What this PR does / why we need it:

When a collateral increases a value, it should be allowed to be withdrawn even if ratio with DFI ratio is still below 50%.

### CODE REVIEW

File: test/functional/feature_loan_vault.py [hunk 1]
```diff
@@ -455,5 +451,42 @@ def run_test(self):
...
+        # BTC doubles in price
+        oracle1_prices = [{"currency": "USD", "tokenAmount": "1@DFI"}, {"currency": "USD", "tokenAmount": "1@TSLA"}, {"currency": "USD", "tokenAmount": "2@BTC"}]
+        timestamp = calendar.timegm(time.gmtime())
+        self.nodes[0].setoracledata(oracle_id1, timestamp, oracle1_prices)
+        self.nodes[0].generate(11)
+
+        # Should be able to withdraw part of BTC
+        self.nodes[0].withdrawfromvault(vaultId4, address, "0.1@BTC")
+        self.nodes[0].generate(1)
+        vault = self.nodes[0].getvault(vaultId4)
```
Comment (Jouzo) [inline test/functional/feature_loan_vault.py:489]:
```suggestion
```

File: src/masternodes/mn_checks.cpp [hunk 2]
```diff
@@ -2505,7 +2505,7 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
                         if (col.nTokenId == DCT_ID{0})
                             totalDFI += col.nValue;
 
-                    if (totalDFI < collateralsLoans.val->totalCollaterals / 2)
+                    if (totalDFI < collateralsLoans.val->totalCollaterals / 2 && obj.amount.nTokenId == DCT_ID{0})
```
Comment (bvbfan) [inline src/masternodes/mn_checks.cpp:2508]: It's a hardfork, guard against GW at line 2503
```
if (obj.amount.nTokenId == DCT_ID{0} || static_cast<int>(height) < consensus.GreatWorldHeight)
```

**Review by bvbfan — APPROVED** [review]

**Review by bvbfan — DISMISSED** [review]

**Review by prasannavl — APPROVED** [review]

File: src/masternodes/mn_checks.cpp [hunk 3]
```diff
@@ -2533,8 +2533,13 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
...
                             totalDFI += col.nValue;
 
-                    if (totalDFI < collateralsLoans.val->totalCollaterals / 2)
-                        return Res::Err("At least 50%% of the vault must be in DFI");
+                    if (static_cast<int>(height) < consensus.FortCanningHillHeight) {
+                        if (totalDFI < collateralsLoans.val->totalCollaterals / 2)
+                            return Res::Err("At least 50%% of the vault must be in DFI");
+                    } else {
+                        if (MultiplyAmounts(totalDFI, 100) < MultiplyAmounts(collateralsLoans.val->totalLoans, scheme->ratio) / 2)
+                            return Res::Err("At least 50%% of the vault must be in DFI");
```
Comment (uzyn) [inline src/masternodes/mn_checks.cpp:2541]:
```suggestion
                            return Res::Err("At least 50% of the collateral must be in DFI");
```

File: src/masternodes/mn_checks.cpp [hunk 4]
```diff
@@ -2639,8 +2644,13 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
                 if (col.nTokenId == DCT_ID{0})
                     totalDFI += col.nValue;
 
-            if (totalDFI < collateralsLoans.val->totalCollaterals / 2)
-                return Res::Err("At least 50%% of the vault must be in DFI when taking a loan");
+            if (static_cast<int>(height) < consensus.FortCanningHillHeight) {
+                if (totalDFI < collateralsLoans.val->totalCollaterals / 2)
+                    return Res::Err("At least 50%% of the vault must be in DFI when taking a loan.");
```
Comment (uzyn) [inline src/masternodes/mn_checks.cpp:2649]:
```suggestion
                    return Res::Err("At least 50% of the collateral must be in DFI when taking a loan.");
```

File: src/masternodes/mn_checks.cpp [hunk 5]
```diff
@@ -2639,8 +2644,13 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
...
+            } else {
+                if (MultiplyAmounts(totalDFI, 100) < MultiplyAmounts(collateralsLoans.val->totalLoans, scheme->ratio) / 2)
+                    return Res::Err("At least 50%% of the vault must be in DFI when taking a loan.");
```
Comment (uzyn) [inline src/masternodes/mn_checks.cpp:2652]:
```suggestion
                    return Res::Err("At least 50% of the collateral must be in DFI when taking a loan.");
```

**Review by uzyn — CHANGES_REQUESTED** [review]
Minor typo. 

For future: Probably good for us to have a centralized error management instead of repeating these.

File: src/masternodes/mn_checks.cpp (hunk 3 above)
Comment (Jouzo) [reply src/masternodes/mn_checks.cpp:2545]: Thanks, I will fix it. we just have to keep the double % otherwise it works as a conversion specifier

**Review by bvbfan — DISMISSED** [review]

**Pull request merged**